crates/polars-ops/src/frame/join/hash_join/single_keys.rs

```diff
@@ -1,53 +1,146 @@
-                let mut offset = 0;
-                for keys in &keys {
-                    let keys = keys.clone().into_iter();
-                    let len = keys.size_hint().1.unwrap() as IdxSize;
+        // FIXME: we wouldn't need this if we changed our interface to split the
+        // input in this function, instead of taking a vec of iterators.
+        let mut per_thread_input_offsets = vec![0; n_partitions];
+        cum_offset = 0;
+        for t in 0..n_threads {
+            per_thread_input_offsets[t] = cum_offset;
+            for p in 0..n_partitions {
+                cum_offset += per_thread_partition_sizes[t][p];
+            }
+        }
 
-                    let mut cnt = 0;
-                    keys.for_each(|k| {
-                        let idx = cnt + offset;
-                        cnt += 1;
+        // Scatter values into partitions.
+        let mut scatter_keys: Vec<T> = Vec::with_capacity(num_keys);
+        let mut scatter_idxs: Vec<IdxSize> = Vec::with_capacity(num_keys);
+        let scatter_keys_ptr = unsafe { SyncPtr::new(scatter_keys.as_mut_ptr()) };
+        let scatter_idxs_ptr = unsafe { SyncPtr::new(scatter_idxs.as_mut_ptr()) };
+        keys.into_par_iter()
+            .enumerate()
+            .for_each(|(t, key_portion)| {
+                let scatter_keys_ptr = scatter_keys_ptr;
+                let scatter_idxs_ptr = scatter_idxs_ptr;
+                let mut partition_offsets =
+                    per_thread_partition_offsets[t * n_partitions..(t + 1) * n_partitions].to_vec();
+                for (i, key) in key_portion.into_iter().enumerate() {
+                    unsafe {
+                        let p = hash_to_partition(key.as_u64(), n_partitions);
+                        let off = partition_offsets.get_unchecked_mut(p);
+                        *scatter_keys_ptr.get().add(*off) = key;
+                        *scatter_idxs_ptr.get().add(*off) =
+                            (per_thread_input_offsets[t] + i) as IdxSize;
+                        *off += 1;
+                    }
+                }
+            });
+        unsafe {
+            scatter_keys.set_len(num_keys);
+            scatter_idxs.set_len(num_keys);
+        }
 
-                        if this_partition(k.as_u64(), partition_no, n_partitions) {
-                            let entry = hash_tbl.entry(k);
+        // Build tables.
+        (0..n_partitions)
+            .into_par_iter()
+            .map(|p| {
+                // Resizing the hash map is very, very expensive. That's why we
+                // adopt a hybrid strategy: we assume an initially small hash
+                // map, which would satisfy a highly skewed relation. If this
+                // fills up we immediately reserve enough for a full cardinality
+                // data set.
+                let partition_range = partition_offsets[p]..partition_offsets[p + 1];
+                let full_size = partition_range.len();
```

Comment:
What do you mean by full cardinality? We don't yet know the cardinality here right?

Reply:
We don't, but we do know that it can't have a higher cardinality than 'completely unique', and at this point we *do* know exactly how many keys are in each partition.